### PR TITLE
make sure to communicate that a failed process ended

### LINKED ISSUE #1360: When launched with "-e", will freeze if ^C instead of quitting
##### Expected Behavior
`qterminal -e cmd` launch a executing cmd. If I ^C before it'd end, the qterminal should quit&close  (as most emulator I know do).

##### Current Behavior
After doing ^C on such a window, instance freeze ever after. It won't return anything, it just will be waiting forever.

##### Possible Solution
Didn't found one other than use another terminal emulator.

##### Steps to Reproduce (for bugs)
1.`qterminal -e sleep 5`
2. input Ctrl+C on it
3. qterminal freeze FOREVER

##### Context
I wasn't even trying to use qterminal specificaly, I'm trying to make a script that'd be environement agnostic using xdg-terminal-exec and qterminal was the one emulator not behaving like the others (which is not a big problem for me, but felt like a glitch).

##### System Information
* QTerminal Version:2.4.0
* Distribution & Version: debian testing
* Kernel: 7.1.3-1 (2026-07-04)
* Qt Version: i don't know !

### PR DESCRIPTION
**Pull request by luebking**
fixes https://github.com/lxqt/qterminal/issues/1360

Ultimately the done() signal needs to be wired up and qterminal prepared for an early exit
In the meantime, forward the termination and print an error on the calling shell

### CODE REVIEW

**Comment by luebking** [comment]
Second iteration: I would not know why to incorporate a notification, just spam an error message and offer to rescue the scroll buffer.

**Comment by luebking** [comment]
~~This causes a coredump on regular exit when closing on of a multi-window process,
Have patch, will update later - don't merge!~~
No more

**Comment by luebking** [comment]
Third iteration doesn't crash in either scenario :)

File: lib/Session.cpp [hunk 1]
```diff
@@ -113,7 +116,7 @@ Session::Session(QObject* parent) :
     connect( _emulation,SIGNAL(useUtf8Request(bool)),_shellProcess,SLOT(setUtf8Mode(bool)) );
 
     connect( _shellProcess,SIGNAL(finished(int,QProcess::ExitStatus)), this, SLOT(done(int,QProcess::ExitStatus)) );
-    // not in kprocess anymore connect( _shellProcess,SIGNAL(done(int)), this, SLOT(done(int)) );
+    connect( _shellProcess, &QProcess::errorOccurred, [=,this](QProcess::ProcessError) { done(-1, QProcess::CrashExit); });
```
Comment (luebking) [inline lib/Session.cpp:119]: This is way to aggressive - it seems QProcess considers SIGUSR1/2 and SIGALRM "erors"

=> Don' t merge for now!

Comment (luebking) [reply lib/Session.cpp:119]: No, is acutally fine - "kill -SIGUSR1 $$" also immediately closes xterm :face_with_diagonal_mouth:

**Comment by tsujan** [comment]
This works as intended according to my simple tests. Can it be merged in its current form? Any edge case or revision you might want to consider?

**Comment by luebking** [comment]
I haven't seen any issues w/ this - it was just while writing the forq™ script that I hadn't trapped SIGUSR1/2 in a long time and was kinda surprised it was closing the TE and I thought this was because of this patch (and somewhat is, w/o it the shell just dies and the terminal becomes dead as in the original bug) - cross testing xterm however revealed that this is not a qterminal problem but the default behavior of at least zsh… tsss.

So that was false alarm and I've currently no concerns reg. the patch.

**Comment by tsujan** [comment]
I think I got "Do you want to save the scrollback buffer?" on two occasions when I was exiting three terminals in three tabs by quickly typing exit. On both occasions it happened in the second tab. But since I was in a hurry, I'm not sure; just pressed Discard and continued.

Don't bother yourself with this comment. I can't reproduce it; it happened randomly.

**Comment by tsujan** [comment]
Oh, I'm lucky:

exit

**Comment by tsujan** [comment]
Now I can reproduce it consistently:

 1. Type a senseless command and press Enter.
 2. Type `exit` and press Enter.

EDIT: BTW, pressing Escape should mean `Discard` but it does nothing.

**Comment by luebking** [comment]
Ah, the previous error is lingering around and the shell exits w/ that.

The string and suggested notification existed and I just connected that but arguably there's no point in annoying the user w/ a dialog if the shell process terminates (cleanly) w/ an error status.
qDebug() / qWarning() might be entirely sufficient - let me sleep over it to come up w/ a (non-posix shell) client that might benefit from saving the screen buffer when exiting w/ an error (but clean signal)

**Comment by stefonarch** [comment]
I got those messages regularly on my testuser at each login, I think it was related to some autostart stuff (backintime?) as I disabled some of them and it didn't happen anymore. Had no time to investigate for good.

**Comment by stefonarch** [comment]
Ok, this:

screen-09-18-08-15-26

happens only if it's not configured.

**Comment by luebking** [comment]
The autostart service tries to run in a TE but fails - either like in the original case case because the binary doesn't exist or because it ran into an error as for @stefonarch

I actually had run into sth. like this - told me there was a (silent) bug in a script that falsely exited w/ an error code after terminating the actual process (fixed the script and stopped worrying)

In that sense the behavior might actually be interesting for "qterminal -e notthedefaultshell" to see the error message of the failing client because there's also no explicit hold switch and it's less likely to run into w/ a false positive of the shell is forwarding the exit status of the last process.

@stefonarch was this more useful (hey, forgot to configure this) or annoying to you?

**Comment by stefonarch** [comment]
it was just annoying and I clicked it away. Only now I read actually what it said ;)

**Comment by luebking** [comment]
So… an in hindsight?
Ie. should we avoid the dialog in this case only for the interactive shell or in general?

(Idk backintime nor how much of an error that is or what it tries to start in a user session context - sounds more like some timeshift/snapshot tool?)

**Comment by tsujan** [comment]
I think that dialog will be annoying to most users if the new version of QTerminal is released with it.

But wasn't the case that I described above a bug?

**Comment by luebking** [comment]
The original bug was qterminal not responding to unclean client exits at all (so SIGINT for the sleep or error 127 because the executable didn't exist)
When addressing that I found those strings and a comment that somebody later might re-introduce the warning, figured it's later and I'm somebody but merely informing the user that their client just crashed was kinda pointless and we could offer them to make the best out of the situation. (This is irrelevant to the bug per se, other TEs will just quit)

As for what to do: I completely agree that for the interactive shell reporting that the user fat-fingered the last command (your case) it's just annoying to raise an alert.
The open question is whether for other clients (@stefonarch backintime autostart and my erroneous script) it's actually helpful because it tells the user that there's a problem w/ the client (even if there's nothing to save in this case)
For me it resulted in killing a (granted: irrelevant) bug in the script; the question for @stefonarch  right now is whether in hindsight (actually reading the dialog and terminal output) that's also the case or it's really just pointless noise.

**Comment by tsujan** [comment]
I mean when I type `exit` at the prompt, the terminal should just accept it. Previous errors shouldn't count, whether with or without showing a dialog.

**Comment by luebking** [comment]
> I completely agree that for the interactive shell reporting that the user fat-fingered the last command (your case) it's just annoying to raise an alert.

The choice is  between shutting this down for the interactive shell only or indiscriminately

**Comment by tsujan** [comment]
> the user fat-fingered the last command (your case)

That's my point: *It's not about fat fingering anything*. `exit` is exit. The previous command shouldn't count; it may be the name of a binary that isn't installed, in which case `bash: xxx: command not found` explains everything.

**Comment by tsujan** [comment]
See the following screenshot. I haven't fat fingered anything; just said "no" to a `[Y/n]` question:

prompt

EDIT: Actually, I now know how to reproduce the issue in two cases: entering a nonexistent command, or answering questions before entering `exit` / pressing Ctrl+d. There may be more cases too.

**Comment by luebking** [comment]
I'll try to stress this as much as I can:

# Your case is and was completely uncontested!

I know what's going on there, why, and why that is a red herring.
I'll even show you a magic trick:
```
false
exit 0
```

---

The other scenario is where the client legit fails, suppose you're creating a desktop entry to ssh into server.com (and tick the box for that to run in the terminal) - but you mistype the ssh command or there's a network or dns issue or whatever.
What will happen is that ssh starts prints an error and then quits.
At this point it might be useful for the user it the terminal doesn't just close but actually stalls and informs the user what happened and esp. also the user gets to see the ssh error.

But that's the scenario @stefonarch ran into and he mentioned that he just found it annoying and didn't even read dialog or the error from backintime.
That's why I'm asking for a revised position after actually reading the error and whether that outweighs the annoyance.

Again: that has nothing to do with what you're facing

# Your case is and was completely uncontested!

I just don't want to draft two patches :)

**Comment by tsujan** [comment]
> I just don't want to draft two patches :)

OK. That "fat fingering" was an alarming sign to me; my bad.

**Comment by stefonarch** [comment]
Just  the fact that I didn't read the message doesn't count. It's actually useful. I just wanted to test something and thought "will look later into it".
So it's actually useful IMO here.

**Comment by luebking** [comment]
https://github.com/lxqt/qtermwidget/pull/678

As for the fat-fingering, try just ctrl+d ;)
